File: tests/csharp/Reference.Tests.CSharp.MSTest/FactorialUnitTest.cs
using Microsoft.Extensions.Logging;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using OpenTracing.Noop;
using OpenTracing.Util;
// ReSharper disable UnusedVariable

namespace Reference.Tests.CSharp.MSTest
{
    /// <summary>
    /// Factorial UnitTest
    /// </summary>
    [TestClass]
    public class FactorialUnitTest
    {
        private ILogger _logger;

        /// <summary>
        /// Initialize test
        /// </summary>
        [TestInitialize]
        public void Init()
        {
            //If no global tracer is registered (not running with scope-run), we register the Noop tracer
            if (!GlobalTracer.IsRegistered())
                GlobalTracer.Register(NoopTracerFactory.Create());

            var loggerFactory = new LoggerFactory()
                // We add the ScopeAgent Logger instrumentation (Automatically added on ASP.NET Core project)
                .AddScopeAgentLogger();

            _logger = loggerFactory.CreateLogger<FactorialUnitTest>();
        }

        /// <summary>
        /// Factorial test
        /// </summary>
        [TestMethod]
        public void FactorialTest()
        {
            Factorial(20);
        }

        /// <summary>
        /// Factorial algorithm
        /// </summary>
        private long Factorial(int number)
        {
            var tracer = GlobalTracer.Instance;
            using (var scope = tracer.BuildSpan("Factorial for: " + number).StartActive())
            {
                long res;
                if (number == 1)
                    res = 1;
                else
                    res = number * Factorial(number - 1);
                _logger.LogInformation("Calculating factorial for: " + number + " = " + res);
                return res;
            }
        }
    }
}

File: tests/fsharp/Reference.Tests.FSharp.MSTest/FactorialUnitTest.fs
namespace Reference.Tests.FSharp.MSTest

open Microsoft.VisualStudio.TestTools.UnitTesting
open OpenTracing.Util
open OpenTracing.Noop
open Microsoft.Extensions.Logging

[<TestClass>]
type FactorialUnitTest () =
    let mutable _logger : ILogger<FactorialUnitTest> = null

    member private this.Factorial(number: int) =
        let tracer = GlobalTracer.Instance
        use scope = tracer.BuildSpan("Factorial for: " + number.ToString()).StartActive()
        let res = if (number = 1) then 1 else number * this.Factorial(number - 1)
        _logger.LogInformation("Calculating factorial for: " + number.ToString() + " = " + res.ToString())
        res

    [<TestInitialize>]
    member this.Init() =
        //If no global tracer is registered (not running with scope-run), we register the Noop tracer
        if (GlobalTracer.IsRegistered() = false) then
            GlobalTracer.Register(NoopTracerFactory.Create())
        let loggerFactory = new LoggerFactory()       
        _logger <- loggerFactory.CreateLogger<FactorialUnitTest>()
        ()
    
    [<TestMethod>]
    member this.FactorialTest () = 
        this.Factorial(20) |> ignore
        ()
